dll/win32/imm32/imm.c

```diff
@@ -2950,40 +3005,134 @@ BOOL WINAPI ImmSetConversionStatus(
     return TRUE;
 }
 
+/***********************************************************************
+ *		ImmLockImeDpi (IMM32.@)
+ */
+PIMEDPI WINAPI ImmLockImeDpi(HKL hKL)
+{
+    PIMEDPI pImeDpi = NULL;
+
+    TRACE("ImmLockImeDpi(%p)\n", hKL);
+
+    RtlEnterCriticalSection(&g_csImeDpi);
+
+    if (g_pImeDpiList)
+    {
+        /* find by hKL */
+        pImeDpi = g_pImeDpiList;
+        do
+        {
+            if (pImeDpi->hKL == hKL) /* found */
+            {
+                /* lock if possible */
+                if (pImeDpi->dwFlags & IMEDPI_FLAG_UNKNOWN)
+                    pImeDpi = NULL;
+                else
+                    ++(pImeDpi->cLockObj);
+                break;
+            }
+            pImeDpi = pImeDpi->pNext;
+        } while (pImeDpi);
+    }
+
+    RtlLeaveCriticalSection(&g_csImeDpi);
+    return pImeDpi;
+}
+
+/***********************************************************************
+ *		ImmUnlockImeDpi (IMM32.@)
+ */
+VOID WINAPI ImmUnlockImeDpi(PIMEDPI pImeDpi)
+{
+    PIMEDPI pEntry, pNext;
+
+    TRACE("ImmUnlockImeDpi(%p)\n", pImeDpi);
+
+    if (pImeDpi == NULL)
+        return;
+
+    RtlEnterCriticalSection(&g_csImeDpi);
+
+    /* unlock */
+    --(pImeDpi->cLockObj);
+    if (pImeDpi->cLockObj != 0)
+    {
+        RtlLeaveCriticalSection(&g_csImeDpi);
+        return;
+    }
+
+    if ((pImeDpi->dwFlags & IMEDPI_FLAG_UNKNOWN) == 0)
+    {
+        if ((pImeDpi->dwFlags & IMEDPI_FLAG_UNKNOWN2) == 0 ||
+            (pImeDpi->dwUnknown1 & 1) == 0)
+        {
+            RtlLeaveCriticalSection(&g_csImeDpi);
+            return;
+        }
+    }
+
+    /* remove from list */
+    pEntry = g_pImeDpiList;
+    if (pEntry == pImeDpi)
+    {
+        g_pImeDpiList = pImeDpi->pNext;
+    }
+    else if (pEntry)
+    {
+        do
+        {
+            pNext = pEntry->pNext;
+            if (pNext == pImeDpi) /* found */
+            {
+                pEntry->pNext = pImeDpi->pNext;
+                break;
+            }
+            pEntry = pNext;
+        } while (pEntry);
+    }
```

Comment:
Suspicious construct, please simplify.
```suggestion
    for (ppEntry = &g_pImeDpiList; *ppEntry; ppEntry = &((*ppEntry)->pNext))
    {
        if (*ppEntry == pImeDpi) /* found */
        {
            *ppEntry = pImeDpi->pNext;
            break;
        }
    }
```

Reply:
OK.